eth/filters/filter_system.go

```diff
@@ -105,8 +112,39 @@ func (sys *FilterSystem) cachedGetLogs(ctx context.Context, blockHash common.Has
 	if logs == nil {
 		return nil, fmt.Errorf("failed to get logs for block #%d (0x%s)", number, blockHash.TerminalString())
 	}
-	sys.logsCache.Add(blockHash, logs)
-	return logs, nil
+	// Database logs are un-derived.
+	// Fill in whatever we can (txHash is inaccessible at this point).
+	flattened := make([]*types.Log, 0)
+	var logIdx uint
+	for i, txLogs := range logs {
+		for _, log := range txLogs {
+			log.BlockHash = blockHash
+			log.BlockNumber = number
+			log.TxIndex = uint(i)
+			log.Index = logIdx
+			logIdx++
+			flattened = append(flattened, log)
+		}
+	}
+	sys.logsCache.Add(blockHash, logCacheElem{logs: flattened})
+	return flattened, nil
+}
+
+func (sys *FilterSystem) cachedGetBody(ctx context.Context, hash common.Hash, number uint64) (*types.Body, error) {
+	cached, ok := sys.logsCache.Get(hash)
```

Comment:
I think we should avoid accessing the cache a second time here. You can just pass the `*logCacheElem` into this function instead.